lib/semantics/resolve-names.cc

```diff
@@ -3796,7 +3793,15 @@ bool DeclarationVisitor::PassesLocalityChecks(
         "Assumed size array '%s' not allowed in a locality-spec"_err_en_US);
     return false;
   }
-  // TODO: Check to see if the name can appear in a variable definition context
+  if (!IsModifiable(symbol, currScope())) {  // C1128
+    SayWithDecl(name, symbol,
+        "'%s' must be able to appear in a variable definition context to "
+        "appear in a locality-spec"_err_en_US);
+    return false;
+  }
+  if (!PassesSharedLocalityChecks(name, symbol)) {
```

Comment:
`PassesSharedLocalityChecks()` has a `IsVariableName()` check in it. There may be a case where the `IsVariableName()` fails, and (because of the reordering) control first goes to one of `IsAllocatable()`, `IsAssumedSizeArray` etc which may not be equipped to handle such a symbol. It may be a risky assumption to make.

Reply:
Currently, there are tests for all of the conditions.  Thus, there are tests where ```IsAllocatable()``` gets called and returns false, and then later ```IsVariableName()``` also gets called and returns false.  For example, in the test resolve56.f90, the test case for ```subroutine s5()``` creates the name ```a``` which is construct associated with an expression.  The ```Symbol``` for ```a``` is neither allocatable or a variable name.  Also note that some names might fail to qualify for being in a locality-spec for multiple reasons.  None of this seems like a problem to me.

I ordered these tests the way that currently appears in the code so that we test for the conditions which produce the most specific error messages first.  For example, a variable which is a coarray of type ```event_type``` gets checked first to see if it's a coarray, which produces a more specific error message than the one caused by being ```event_type``` since the latter fails because with a message saying that the variable name must be able to appear in a variable definition context.

Reply:
My theory was that there could be symbols which are not handled in std::visits that exist in one of the other checks, which would have been caught by `IsVariableName()`. However I was unsuccessful in coming up with a case to test out this theory.